dask/array/core.py

```diff
@@ -1002,6 +1002,22 @@ def __array__(self, dtype=None, **kwargs):
             x = np.array(x)
         return x
 
+    def __array_function__(self, func, types, args, kwargs):
+        import dask.array as module
+        for submodule in func.__module__.split('.')[1:]:
+            try:
+                module = getattr(module, submodule)
+            except AttributeError:
+                return NotImplemented
+        if not hasattr(module, func.__name__):
+            return NotImplemented
+        da_func = getattr(module, func.__name__)
+        if da_func is func:
+            return NotImplemented
+        if not all(issubclass(t, Array) for t in types):
+            return NotImplemented
```

Comment:
The `__array_function__` protocol guarantees that it will only get called if `dask.array.Array` is in types. So if that's all you wanted to check, you could drop this entirely.

The problem is that you want to support operations involving some but not all other array types. For example, `xarray.DataArray` wraps dask arrays, not the other way around. So `dask.Array.__array_function__` should return `NotImplemented` in that case, leaving it up `xarray.DataArray.__array_function__` to define the operation.

We basically need some protocol or registration system to mark arrays as "can be coerced into dask array", e.g., either
```
sparse.Array.__dask_chunk_compatibile__ = True
```
or
```
dask.array.register_chunk_type(sparse.Array)
```

Reply:
Hrm, that's an interesting problem.  Maybe we try to turn all of the args into dask arrays with `da.asarray` and, if that works, proceed?

I'm not sure that this would produce the correct behavior for `xarray.DataArray` though.  I genuinely don't know what the right behavior is there.

Reply:
I actually think my idea for mixins can help with this. Specifically, we could check the specific mixins that Dask absolutely requires, I’m assuming that’d be the indexing mixin.

Reply:
> Hrm, that's an interesting problem. Maybe we try to turn all of the args into dask arrays with `da.asarray` and, if that works, proceed?

I actually like this idea for its simplicity.

> If we want to be completely general, we could use `hasattr(t, ‘__array_function__’)`. But this would also include Pandas dataframes at some point I’m assuming...

I also like this idea in general, but I think then it would be best before to agree on all mixins we expect to have/support. Plus, if I understood correctly, it feels like this will require much more type checking, which may become complicated to handle.

Reply:
From my perspective as an xarray maintainer, casting an xarray.DataArray into a dask.array should do one of two things:
- raise TypeError, or
- pull out .data, and cast that to a Dask array if it isn't already

We don't want xarray.DataArray inside Dask arrays, because xarray objects have intentionally different behavior from NumPy arrays, so Dask array operations would not be well defined.

Reply:
I agree with the `TypeError`. The whole point of the protocol was to avoid these kinds of project-specific dependencies... I'm comfortable with @mrocklin's idea of casting to a `dask.Array` provided that it performs the correct checks and actually returns a Dask array (I'm assuming this would have to be done within functions anyway), although I guess at that point we're kicking the can down the road, because similar issues arise within the constructor.

Reply:
Also, since XArray is a container of arrays like Dask and expects a NumPy-like API, it'll face similar issues.

Reply:
A related issue is maintaining a well defined type casting hierarchy: https://www.numpy.org/neps/nep-0013-ufunc-overrides.html#type-casting-hierarchy

Relying on a generic mixin would be fine, but ideally it would be a mixin that Dask.array itself does not support. Otherwise you'll end up in situations where type casting is not commutative (e.g., `type(np.stack([a, b])) != type(np.stack([b, a]))`), which we would really like to avoid. I guess Dask could safely coerce anything that implements InMemory and NumpyDuckArray mixins?

A special attribute or method seems a little easier since that's something that can be monkey patched in if desired (vs injecting another super class, which I'm sure is possible but seems tricky). I like the idea of encouraging library authors to explicitly "opt in" to Dask compatibility but not strictly requiring it.

Reply:
The way I see it is there are two ways to classify things:

- Implements the NumPy API or not (let's ignore for now what APIs are implemented)
- Contains NumPy-like arrays or not.

Common examples:
- CuPy (contains-> no, implements->yes)
- Sparse (contains-> no (but could contain CuPy, maybe... A bit tricky because more than one array), implements->yes)
- Dask (contains-> yes, implements->yes)
- XArray (contains-> yes, implements->no)

We only run into the problem you're suggesting when there is more than one library where the answer to both questions is yes. And it looks like that might become the case.

So here's what I suggest: APIs that contain other arrays should have a clear protocol of extracting the underlying arrays which then _will not_ be containers.

Reply:
For what it's worth I'm thinking of a similar "opt in" approach with xarray. There's no intrinsic reason for preferring `UnitsArray` wrapping `AxisLabelsArray` or the other way around, but either way it will drive our users nuts if they can't rely on the ordering of the array wrappers.

The virtue of a dask specific special method or attribute is that it allows developers of third party libraries to make an explicit decision about how they want to interact with Dask rather than making the decision for them. Refusing the temptation to guess in the face of ambiguity is the polite thing to do.

Reply:
UnitsArray, e.g., as implemented in Pint, is probably a good example of another array type satisfies all of these:
1. implements NumPy's API
2. contains other duck arrays
3. could be usefully wrap or be wrapped with/by dask arrays

Reply:
> I guess Dask could safely coerce anything that implements `InMemory` and `NumpyDuckArray` mixins?

I was thinking of three classes of (NumPy-Specific) mixins, aimed at implementing what I see as the three major protocols:

- `NumPyUfuncMixin`
- `NumPyIndexingMixin` (for NumPy-like indexing, so `XArray` and `pandas` would skip this).
- `NumPyArrayFunctionMixin` (for implementing array methods via `__array_function__`, such as `sum` and `mean`).

I see the use of your `InMemory` mixin, but I would rename it to `InCore`, as GPU computations don't take place in main memory. :)

For now, these would be provisional, just like `__array_function__`. The hope is that these would be adopted and specific checking code would go away.

Reply:
I find the mixin idea challenging for two reasons:

1.  I think that agreeing on a few major protocols will be difficult.  Different groups will have different ideas of what these protocols are.  I think that it will be difficult to reach any sort of consensus here 
2.  I think that getting upstream projects to inherit from NumPy classes will be difficult

Reply:
Then how about a "numpy/pydata community repo" containing these mixins? I would be happy to maintain one. (Not in the main NumPy codebase)? The reason I like the idea of mixins is it allows "duck-array" like objects to easily implement NumPy-like functionality without too much effort.

We can also have protocols:

```
class DuckArray:
    __array_indexing__ = True
    __array_in_core__ = True
```

Reply:
I think that that would only help to address the second point above.  I'm not sure that it would solve it though.  As a worst case scenario I suspect that a project like TensorFlow would not inherit from those classes.  We could get NumPy, Dask, Sparse, and Xarray on board but I'm not sure about any of the others out there.

I personally think that duck typing is more likely to work across projects than explicit typing.

Mixins may still be a good idea for the reason you mention, you can easily implement a lot of NumPy-like functionality without too much effort, but I don't think that they're a good choice to test for Numpy-ness.

Reply:
I do agree that it's hard to get traction for mixins or add dependencies, but adding a protocol isn't hard.

While you're right that TensorFlow doesn't plan on implementing any of these protocols, it has always historically done its own thing, and it has its own distributed system. I suspect that with Google's historical "rewrite the world and do it better" philosophy, it'll be hard to get them to play along anyway.

In contrast, anyone specifically aiming to be a NumPy duck array won't have a problem implementing protocols, and that is what I'm planning to do, to allow mixins to implement methods and so forth based on the implementation of protocols.

Reply:
For the time being, I've used @shoyer and @mrocklin suggestions replacing only that last condition in `__array_function__` to:

```python
if not all(issubclass(t, Array) for t in types):
    return NotImplemented
```

It seems to me that the mixin discussion has grown enough that we should move it to a Dask issue, discussed there and later review the current code. Anybody disagrees with my proposal?

Reply:
Since nobody opposed, I moved the discussion to https://github.com/dask/dask/issues/4583.

Reply:
I think that with @shoyer 's statements below we can drop this conditional entirely.  It should always pass

>  The __array_function__ protocol guarantees that it will only get called if dask.array.Array is in types. So if that's all you wanted to check, you could drop this entirely.